File: src/main/java/club/mindtech/mindbot/commands/CommandHelp.java
package club.mindtech.mindbot.commands;

import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.events.interaction.SlashCommandEvent;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
import net.dv8tion.jda.api.interactions.commands.OptionMapping;
import net.dv8tion.jda.api.interactions.commands.OptionType;
import net.dv8tion.jda.api.interactions.commands.build.CommandData;

import java.util.List;
import java.util.Locale;

import static club.mindtech.mindbot.helpers.StringHelper.*;

public class CommandHelp extends BaseCommand {

    public CommandHelp() {
        super("help", "Shows information about other commands", "help [command]", "h");
    }

    @Override
    public boolean isSlashCommand() {
        return true;
    }

    @Override
    public CommandData getCommandData() {
        CommandData data = super.getCommandData();
        data.addOption(OptionType.STRING, "command", "Command name the help should be shown for", false);
        return data;
    }

    @Override
    public void onCommand(MessageReceivedEvent event, List<String> args) {
        super.onCommand(event, args);
        BaseCommand embedFor = this;

        if (args.size() > 0 && Commands.isCommand(args.get(0))) {
            embedFor = Commands.getCommand(lower(args.get(0)));
        }

        event.getMessage()
             .replyEmbeds(getHelpEmbed(event.getAuthor(), embedFor))
             .queue();
    }

    @Override
    public void onSlashCommand(SlashCommandEvent event) {
        super.onSlashCommand(event);

        BaseCommand embedFor = this;
        OptionMapping mapping = event.getOption("command");

        if (mapping != null) {
            String commandName = lower(mapping.getAsString());
            if (Commands.isCommand(commandName)) {
                embedFor = Commands.getCommand(commandName);
            }
        }

        event.deferReply()
             .addEmbeds(getHelpEmbed(event.getUser(), embedFor))
             .queue();
    }

    private static MessageEmbed getHelpEmbed(User author, BaseCommand command) {
        EmbedBuilder embed = new EmbedBuilder()
                .setTitle("Command: " + command.getName())
                .setDescription(command.getDescription())
                .addField("Usage:", command.getUsage(), true)
                .addField("Aliases:", stringify(command.getAliases()), true)
                .setColor(0x1dd1a1)
                .setFooter("Requested by " + author.getName());

        return embed.build();
    }

    private static String lower(String s) {
        return s.toLowerCase(Locale.ROOT);
    }
}

File: bot/src/main/kotlin/dev/crec/hawksbill/commands/CommandHelp.kt
package dev.crec.hawksbill.commands

import dev.minn.jda.ktx.interactions.commands.option
import dev.minn.jda.ktx.messages.EmbedBuilder
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.events.interaction.command.CommandAutoCompleteInteractionEvent
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.interactions.commands.build.SlashCommandData

class CommandHelp : BaseCommand("help", "Shows information about other commands", "help [command]") {

    override fun getCommandData(): SlashCommandData {
        return super.getCommandData {
            option<String>(
                name = "command",
                description = "Command name the help should be shown for",
                required = false,
                autocomplete = true
            )
        }
    }

    override fun onSlashCommand(event: SlashCommandInteractionEvent) {
        val commandName: String? = event.getOption("command")?.asString?.lowercase()
        event.deferReply().addEmbeds(getHelpEmbed(event.user, commandName)).queue()
    }

    override fun onAutoComplete(event: CommandAutoCompleteInteractionEvent) {
        val userInput = event.focusedOption.value
        val matchingCommands = getRegisteredCommands().keys.filter { it.contains(userInput) }
        event.replyChoiceStrings(matchingCommands).queue()
    }

    private fun getHelpEmbed(author: User, commandName: String?): MessageEmbed {
        val command: BaseCommand? = getCommand(commandName)

        val titleString = if (commandName == null) {
            "All Available commands"
        } else if (command == null) {
            "Command not found"
        } else {
            "Help for command: ${command.name}"
        }

        val builder = EmbedBuilder {
            title = titleString
            color = 0x1dd1a1
            description = command?.description ?: availableCommands()
            footer {
                name = "Requested by ${author.name}"
            }
        }

        command?.let {
            builder.field {
                name = "Usage:"
                value = command.usage
                inline = true
            }
        }

        return builder.build()
    }

    private fun availableCommands(): String {
        return """
            Available commands:
            ${getRegisteredCommands().keys.joinToString("\n") { "**»** $it" }}
            """
    }
}
